java-checks/src/main/java/org/sonar/java/checks/DataStoredInSessionCheck.java

```diff
@@ -0,0 +1,112 @@
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ReassignmentFinder;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodInvocationMatcherCollection;
+import org.sonar.java.checks.methods.MethodMatcher;
+import org.sonar.java.checks.methods.NameCriteria;
+import org.sonar.java.checks.methods.TypeCriteria;
+import org.sonar.java.syntaxtoken.FirstSyntaxTokenFinder;
+import org.sonar.java.syntaxtoken.LastSyntaxTokenFinder;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S3318",
+  name = "Untrusted data should not be stored in sessions",
+  priority = Priority.CRITICAL,
+  tags = {"cwe", "security"})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.INPUT_VALIDATION_AND_REPRESENTATION)
+@SqaleConstantRemediation("20min")
+public class DataStoredInSessionCheck extends AbstractMethodDetection {
+
+  private static final MethodInvocationMatcherCollection REQUEST_OR_COOKIE_DATA_RETRIEVAL = MethodInvocationMatcherCollection.create(
+    MethodMatcher.create().typeDefinition("javax.servlet.http.Cookie").name(NameCriteria.startsWith("get")).withNoParameterConstraint(),
+    MethodMatcher.create().callSite(TypeCriteria.is("javax.servlet.http.HttpServletRequest")).name(NameCriteria.startsWith("get")).withNoParameterConstraint(),
+    MethodMatcher.create().callSite(TypeCriteria.is("javax.servlet.http.HttpServletRequest")).name(NameCriteria.startsWith("get")).addParameter(TypeCriteria.anyType()));
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodMatcher.create().typeDefinition("javax.servlet.http.HttpSession").name("setAttribute").withNoParameterConstraint());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree expressionToEvaluate = mit.arguments().get(1);
+    if (expressionToEvaluate.is(Tree.Kind.IDENTIFIER)) {
+      Symbol variable = ((IdentifierTree) expressionToEvaluate).symbol();
+      ExpressionTree lastAssignmentOrDeclaration = ReassignmentFinder.getClosestReassignmentOrDeclarationExpression(mit, variable);
+      if (lastAssignmentOrDeclaration != null && !usedBetween(variable, lastAssignmentOrDeclaration, mit)) {
+        expressionToEvaluate = lastAssignmentOrDeclaration;
+      }
+    }
+
+    if (isRequestOrCookieDataRetrieval(expressionToEvaluate)) {
+      addIssue(mit, "Make sure the user is authenticated before this data is stored in the session.");
```

Comment:
@m-g-sonar you can try the new api: `reportIssue(mit.methodSelect(), "Make sure the user is authenticated before this data is stored in the session.");`